File: androidApp/src/main/kotlin/com/example/app/sdkwrapper/NativePaymentSdk.kt
package com.example.app.sdkwrapper

import android.app.Activity
import com.android.billingclient.api.*
import com.example.paymentsdk.PaymentSdk
import com.example.paymentsdk.models.Product
import com.example.paymentsdk.models.PurchaseResult
import com.example.paymentsdk.models.Transaction
import com.example.paymentsdk.models.TransactionStatus
import kotlinx.coroutines.CancellableContinuation
import kotlinx.coroutines.suspendCancellableCoroutine
import java.util.concurrent.atomic.AtomicReference
import kotlin.coroutines.resume

/**
 * Android implementation of [PaymentSdk] wrapping
 * Google Play Billing Library.
 *
 * Lives in the androidApp layer because Google Play
 * Billing is an Android-only SDK.
 */
class NativePaymentSdk(
    private val activity: Activity
) : PaymentSdk {

    private val pendingPurchaseContinuation =
        AtomicReference<CancellableContinuation<PurchaseResult>?>(null)

    private val billingClient = BillingClient.newBuilder(activity)
        .setListener { billingResult, purchases ->

            val continuation =
                pendingPurchaseContinuation.getAndSet(null)

            when (billingResult.responseCode) {

                BillingClient.BillingResponseCode.OK -> {
                    val orderId =
                        purchases?.firstOrNull()?.orderId ?: ""
                    continuation?.resume(
                        PurchaseResult.Success(orderId)
                    )
                }

                BillingClient.BillingResponseCode.USER_CANCELED -> {
                    continuation?.resume(
                        PurchaseResult.UserCanceled
                    )
                }

                else -> {
                    continuation?.resume(
                        PurchaseResult.Error(
                            billingResult.responseCode,
                            billingResult.debugMessage
                        )
                    )
                }
            }
        }
        .enablePendingPurchases()
        .build()

    // ---- Connection ----

    suspend fun connect() = suspendCancellableCoroutine { cont ->
        billingClient.startConnection(
            object : BillingClientStateListener {
                override fun onBillingSetupFinished(
                    result: BillingResult
                ) {
                    cont.resume(Unit)
                }
                override fun onBillingServiceDisconnected() {}
            }
        )
    }

    private suspend fun ensureConnected() {
        if (!billingClient.isReady) connect()
    }

    suspend fun syncMissedPurchases() {
        ensureConnected()
        val result = billingClient.queryPurchasesAsync(
            QueryPurchasesParams.newBuilder()
                .setProductType(BillingClient.ProductType.SUBS)
                .build()
        )
        // Process un-acknowledged purchases...
    }

    // ---- PaymentSdk Interface ----

    override suspend fun getProducts(
        productIds: List<String>
    ): List<Product> {

        ensureConnected()

        val productList = productIds.map { id ->
            QueryProductDetailsParams.Product.newBuilder()
                .setProductId(id)
                .setProductType(BillingClient.ProductType.SUBS)
                .build()
        }

        val queryParams = QueryProductDetailsParams.newBuilder()
            .setProductList(productList)
            .build()

        val (billingResult, detailsList) =
            billingClient.queryProductDetails(queryParams)

        if (billingResult.responseCode !=
            BillingClient.BillingResponseCode.OK
        ) {
            return emptyList()
        }

        return detailsList?.map { details ->
            val phase = details.subscriptionOfferDetails
                ?.firstOrNull()
                ?.pricingPhases
                ?.pricingPhaseList
                ?.firstOrNull()

            Product(
                productId = details.productId,
                title = details.name,
                description = details.description,
                formattedPrice = phase?.formattedPrice ?: "",
                price = (phase?.priceAmountMicros
                    ?: 0) / 1_000_000.0,
                currencyCode = phase?.priceCurrencyCode ?: ""
            )
        } ?: emptyList()
    }

    override suspend fun purchase(
        product: Product
    ): PurchaseResult {

        ensureConnected()

        return suspendCancellableCoroutine { continuation ->

            pendingPurchaseContinuation.set(continuation)

            continuation.invokeOnCancellation {
                pendingPurchaseContinuation.compareAndSet(
                    continuation, null
                )
            }

            // Look up cached ProductDetails by productId
            // and launch billing flow...
            // billingClient.launchBillingFlow(activity, params)
        }
    }

    /**
     * Queries the store for the transaction and, if it is in
     * the `PURCHASED` state and not yet acknowledged, calls
     * `acknowledgePurchase` (subs / non-consumable) or
     * `consumeAsync` (consumables) before returning.
     *
     * Google Play auto-refunds purchases that are not
     * acknowledged within 3 days, so this must run on every
     * successful purchase.
     *
     * Backend stays in sync via Play Real-time Developer
     * Notifications (Google → backend), and the backend pushes
     * updates to the app out-of-band (FCM) — no HTTP call from
     * the SDK is needed.
     */
    override suspend fun getTransactionResult(
        transactionId: String
    ): Transaction {

        ensureConnected()

        val result = billingClient.queryPurchasesAsync(
            QueryPurchasesParams.newBuilder()
                .setProductType(BillingClient.ProductType.SUBS)
                .build()
        )

        val purchase = result.purchasesList.firstOrNull {
            it.orderId == transactionId
        }

        if (purchase != null &&
            purchase.purchaseState == Purchase.PurchaseState.PURCHASED &&
            !purchase.isAcknowledged
        ) {
            // acknowledgePurchase() for subs / non-consumables,
            // or consumeAsync() for consumables.
            val ackParams = AcknowledgePurchaseParams.newBuilder()
                .setPurchaseToken(purchase.purchaseToken)
                .build()
            billingClient.acknowledgePurchase(ackParams)
        }

        return Transaction(
            transactionId = transactionId,
            productId = purchase?.products?.first() ?: "",
            receiptToken = purchase?.purchaseToken ?: "",
            status = when (purchase?.purchaseState) {
                Purchase.PurchaseState.PENDING ->
                    TransactionStatus.PENDING
                Purchase.PurchaseState.PURCHASED ->
                    TransactionStatus.COMPLETED
                else ->
                    TransactionStatus.FAILED
            },
            purchasedAt = purchase?.purchaseTime ?: 0
        )
    }
}

File: iosApp/SdkWrapper/NativePaymentSdk.swift
import StoreKit
import Foundation
import shared

/// iOS implementation of PaymentSdk wrapping StoreKit 2.
///
/// Lives in the iosApp layer because StoreKit 2 is
/// Swift-only (no Kotlin/Native interop).
class NativePaymentSdk: PaymentSdk {

    // MARK: - Get Products

    func getProducts(
        productIds: [String]?
    ) async throws -> [Product] {

        // StoreKit 2 requires explicit product ids; there is
        // no "list all" API. Caller must supply them.
        guard let ids = productIds, !ids.isEmpty else {
            return []
        }

        let storeProducts = try await StoreKit.Product.products(
            for: ids
        )

        return storeProducts.map { p in
            Product(
                productId: p.id,
                title: p.displayName,
                description: p.description,
                formattedPrice: p.displayPrice,
                price: NSDecimalNumber(
                    decimal: p.price
                ).doubleValue,
                currencyCode: p.priceFormatStyle
                    .currencyCode ?? ""
            )
        }
    }

    // MARK: - Purchase

    func purchase(
        product: Product
    ) async throws -> PurchaseResult {

        guard let storeProduct = try await
            StoreKit.Product.products(
                for: [product.productId]
            ).first
        else {
            return PurchaseResult.Error(
                code: -1,
                message: "Product not found"
            )
        }

        let result = try await storeProduct.purchase()

        switch result {

        case .success(let verification):
            guard case .verified(let transaction)
                = verification else {
                return PurchaseResult.Error(
                    code: -1,
                    message: "Verification failed"
                )
            }
            return PurchaseResult.Success(
                transactionId: String(transaction.id)
            )

        case .userCancelled:
            return PurchaseResult.UserCanceled.shared

        case .pending:
            return PurchaseResult.Success(
                transactionId: "pending"
            )

        @unknown default:
            return PurchaseResult.Error(
                code: -1,
                message: "Unknown"
            )
        }
    }

    // MARK: - Get Transaction Result

    /// Returns the verified transaction from StoreKit and calls
    /// `transaction.finish()` so StoreKit stops re-delivering
    /// it on subsequent launches.
    ///
    /// Backend stays in sync via App Store Server Notifications
    /// V2 (Apple → backend), and the backend pushes updates to
    /// the app out-of-band (APNs / silent push) — no HTTP call
    /// from the SDK is needed.
    func getTransactionResult(
        transactionId: String
    ) async throws -> Transaction {

        guard let txId = UInt64(transactionId) else {
            return Transaction(
                transactionId: transactionId,
                productId: "",
                receiptToken: "",
                status: .failed,
                purchasedAt: 0
            )
        }

        for await result in StoreKit.Transaction.all {
            guard case .verified(let tx) = result,
                  tx.id == txId else { continue }

            await tx.finish()

            return Transaction(
                transactionId: String(tx.id),
                productId: tx.productID,
                receiptToken: tx.jwsRepresentation,
                status: .completed,
                purchasedAt: Int64(
                    tx.purchaseDate
                        .timeIntervalSince1970 * 1000
                )
            )
        }

        return Transaction(
            transactionId: transactionId,
            productId: "",
            receiptToken: "",
            status: .failed,
            purchasedAt: 0
        )
    }
}
